tools/c7n_azure/c7n_azure/resources/access_control.py

```diff
@@ -82,3 +127,85 @@ class UserRole(RelatedResourceFilter):
 
     RelatedResource = "c7n_azure.resources.access_control.RoleDefinition"
     RelatedIdsExpression = "properties.roleDefinitionId"
+
+
+@RoleAssignment.filter_registry.register('resource-access')
+class ResourceAccessFilter(Filter):
+    """Filters role assignments that have access to a certain
+    type of azure resource.
+
+    :Example:
+
+        .. code-block:: yaml
+
+            policies:
+               - name: assignments-by-azure-resource
+                 resource: azure.roleassignment
+                 filters:
+                    - type: resource-access
+                      relatedResource: azure.vm
+    """
+
+    schema = type_schema(
+        'resource-access',
+        relatedResource={'type': 'string'},
+        required=['relatedResource']
+    )
+
+    def __init__(self, data, manager=None):
+        super(ResourceAccessFilter, self).__init__(data, manager)
+        resource_type = self.data['relatedResource']
+        self.factory = Azure.resources.get(
+            resource_type.rsplit('.', 1)[-1])
+
+    def get_related(self):
+        ctx = ExecutionContext(local_session(Session), self.data, Config.empty())
+
+        manager = self.factory(ctx, self.data)
+        return manager.source.get_resources(None)
+
+    def process(self, resources, event=None):
+        related_resources = self.get_related()
+        s = Session()
+        client = s.client('azure.mgmt.authorization.AuthorizationManagementClient')
+        assignment_ids = []
+        for resource in related_resources:
+            results = client.role_assignments.list_for_resource(
+                ResourceIdParser.get_resource_group(resource['id']),
+                ResourceIdParser.get_namespace(resource['id']),
+                '',
```

Comment:
The empty string represents the parent resource. If it's likely I can't make the assumption that resources don't have a parent then I'll have to add some more robust ID parsing 👍